components/applications-service/pkg/generator/runner.go

```diff
@@ -0,0 +1,253 @@
+
+	fmt.Fprintf(&b, "- Simulated supervisors: %d\n", s.Count)
+	fmt.Fprintf(&b, "- Services per supervisor: %d\n", len(s.MessagePrototypes))
+
+	fmt.Fprintln(&b, "")
+
+	for _, m := range s.MessagePrototypes {
+		fmt.Fprintln(&b, m.PrettyStr())
+	}
+
+	return b.String()
+}
+
+func (s *SupervisorGroup) Run(cfg *RunnerConfig, stats *RunnerStatsKeeper) {
+	for n := int32(0); n < s.Count; n++ {
+		log.WithFields(log.Fields{"name": s.Name, "index": n}).Info("spawning supervisor")
+		go s.SpawnSup(n, cfg, stats)
+	}
+}
+
+func (s *SupervisorGroup) SpawnSup(n int32, cfg *RunnerConfig, stats *RunnerStatsKeeper) error {
+	uuid, err := uuid.NewV4()
+	if err != nil {
+		log.WithError(err).Error("Failed to make a V4 UUID")
+		return err
+	}
+
+	sup := &SupSim{
+		Name:              s.Name,
+		UUID:              uuid,
+		Cfg:               cfg,
+		MessagePrototypes: s.MessagePrototypes,
+		Stats:             stats,
+	}
+
+	err = sup.Run()
+	if err != nil {
+		fmt.Printf("Supervisor failed to run: %s\n", err)
+		return err
+	}
+
+	return nil
+}
+
+type SupSim struct {
+	Name              string
+	Idx               int32
+	Cfg               *RunnerConfig
+	UUID              uuid.UUID
+	MessagePrototypes []*MessagePrototype
+	Stats             *RunnerStatsKeeper
+	nc                *nats.NatsClient
+}
+
+func (s *SupSim) Run() error {
+	s.Stats.SupStarted()
+	defer s.Stats.SupDied()
+
+	url := fmt.Sprintf("nats://%s@%s:4222", s.Cfg.AuthToken, s.Cfg.Host)
+	cluster := "event-service"
+	client := fmt.Sprintf("load-generator-%s", s.UUID)
+	durable := client
+	subject := "habitat"
+
+	s.nc = nats.NewExternalClient(url, cluster, client, durable, subject)
+	s.nc.InsecureSkipVerify = true
+	err := s.nc.Connect()
+	if err != nil {
+		return err
+	}
+	defer s.nc.Close()
+	// run loadgen loop
+
+	log.WithFields(log.Fields{"name": s.Name, "index": s.Idx, "uuid": s.UUID.String()}).Info("starting supervisor")
+
+	// the ticker always waits its full time before publishing. Also we wish to
+	// splay out the simulated sups for a more even distribution of messages. So
+	// we get a random splay, publish once, then loop.
+	splay := rand.Int31n(int32(s.Cfg.Tick))
+	time.Sleep(time.Duration(splay) * time.Second)
+	s.PublishAll()
+
+	ticker := time.NewTicker(time.Duration(s.Cfg.Tick) * time.Second)
+	for _ = range ticker.C {
+		s.PublishAll()
+	}
+	return nil
+}
+
+func (s *SupSim) PublishAll() error {
+	log.WithFields(log.Fields{
+		"name":          s.Name,
+		"index":         s.Idx,
+		"uuid":          s.UUID.String(),
+		"service_count": len(s.MessagePrototypes),
+	}).Info("publishing messages")
+
+	for _, m := range s.MessagePrototypes {
+
+		msg := m.CreateMessage(s.UUID.String())
+
+		log.WithFields(log.Fields{
+			"name":          s.Name,
+			"index":         s.Idx,
+			"uuid":          s.UUID.String(),
+			"service_count": len(s.MessagePrototypes),
+			"pkg_fqid":      fmt.Sprintf("%s/%s/%s/%s", m.Origin, m.PkgName, versionNumber, m.Release),
```

Comment:
Would be nice to have a function attached to the `MessagePrototype` struct that generates this full release string, just like: 

https://github.com/chef/automate/blob/master/components/applications-service/pkg/storage/client.go#L46-L49

Reply:
done!